File: cli/tool_display.py
"""Tool execution status display panels."""

from __future__ import annotations

import json

from rich.console import Console
from rich.markup import escape as markup_escape
from rich.panel import Panel
from rich.text import Text

from cli.spinner import Spinner


class ToolDisplay:
    """Formatted panels for tool call start and result display."""

    TOOL_ICONS = {
        "bash": "\u26a1",
        "file_read": "\U0001f4c4",
        "file_write": "\u270f\ufe0f",
        "file_edit": "\U0001f4dd",
        "web_search": "\U0001f50e",
        "calculator": "\U0001f9ee",
        "weather": "\U0001f324\ufe0f",
    }

    def __init__(self, console: Console) -> None:
        self.console = console
        self._active_spinner: Spinner | None = None

    def show_tool_start(self, tool_name: str, tool_input: dict) -> None:
        """Display a formatted panel when a tool call begins."""
        safe_name = markup_escape(str(tool_name))
        icon = self.TOOL_ICONS.get(tool_name, "\U0001f527")
        detail = self._format_detail(tool_name, tool_input)
        # Loading indicator: ⏺ Calling tools... — shows tool is running
        loading = Text()
        loading.append("⏺ ", style="dim")
        loading.append("Calling tools...", style="dim")
        self.console.print(loading)
        panel = Panel(
            detail,
            title="[tool.name]{} {}[/]".format(icon, safe_name),
            title_align="left",
            border_style="tool.border",
            padding=(0, 1),
        )
        self.console.print(panel)

    def show_tool_result(
        self,
        tool_name: str,
        result: str,
        elapsed_ms: float,
        is_error: bool = False,
    ) -> None:
        """Display a one-line summary when a tool call completes."""
        # 停止之前的 thinking spinner（如果有）
        self.stop_thinking()
        safe_name = markup_escape(str(tool_name))
        icon = self.TOOL_ICONS.get(tool_name, "\U0001f527")
        status_style = "red" if is_error else "green"
        status_icon = "\u2717" if is_error else "\u2713"
        elapsed_str = "{:.0f}ms".format(elapsed_ms)

        preview = result[:200].replace("\n", " ")
        if len(result) > 200:
            preview += "\u2026"

        text = Text()
        text.append("\u23bf ", style="dim")
        text.append("{} ".format(icon), style="dim")
        text.append(safe_name, style="tool.name")
        text.append("  {}".format(status_icon), style=status_style)
        text.append("  {}".format(elapsed_str), style="dim cyan")
        if is_error:
            text.append("  {}".format(preview), style="red")
        self.console.print(text)
        # 工具执行完毕，启动 thinking spinner 等待 LLM 响应
        self.start_thinking()

    def start_thinking(self) -> None:
        """启动 thinking spinner，在 LLM 思考期间显示动画。"""
        if self._active_spinner is None:
            self._active_spinner = Spinner(self.console, label="Thinking...")
            self._active_spinner.start()

    def stop_thinking(self) -> None:
        """停止 thinking spinner。"""
        if self._active_spinner is not None:
            self._active_spinner.stop()
            self._active_spinner = None

    def _format_detail(self, tool_name: str, tool_input: dict) -> str:
        """Return tool-specific detail string for the panel body."""
        if not isinstance(tool_input, dict):
            try:
                preview = json.dumps(tool_input, ensure_ascii=False)[:150]
            except (TypeError, ValueError):
                preview = repr(tool_input)[:150]
            return markup_escape(preview)

        if tool_name == "bash":
            cmd = markup_escape(str(tool_input.get("command", "")))
            return "$ {}".format(cmd)
        elif tool_name == "file_read":
            path = markup_escape(str(tool_input.get("path", "")))
            return "\U0001f4c4 Reading {}".format(path)
        elif tool_name == "file_write":
            path = markup_escape(str(tool_input.get("path", "")))
            content = str(tool_input.get("content", ""))
            lines = content.count("\n") + 1
            return "\u270f\ufe0f Writing {} ({} lines)".format(path, lines)
        elif tool_name == "file_edit":
            path = markup_escape(str(tool_input.get("path", "")))
            return "\U0001f4dd Editing {}".format(path)
        elif tool_name == "web_search":
            query = markup_escape(str(tool_input.get("query", "")))
            return "\U0001f50e Searching: {}".format(query)
        else:
            try:
                preview = json.dumps(tool_input, ensure_ascii=False)[:150]
            except (TypeError, ValueError):
                preview = repr(tool_input)[:150]
            return markup_escape(preview)

    def cleanup(self) -> None:
        """安全清理：异常路径调用，确保 spinner 不泄漏"""
        self.stop_thinking()

File: src/cli/tool-display.ts
import chalk from 'chalk'
import { Spinner } from './spinner.js'

/** 工具图标映射 */
const TOOL_ICONS: Record<string, string> = {
  bash: '⚡',
  file_read: '📄',
  file_write: '✏️',
  file_edit: '📝',
  web_search: '🔍',
  calculator: '🧮',
  weather: '🌤️',
}

/**
 * ToolDisplay - 工具调用面板显示
 * 展示工具调用的开始、结果，以及中间的 thinking spinner
 */
export class ToolDisplay {
  private activeSpinner: Spinner | null = null

  /** 显示工具调用开始面板 */
  showToolStart(toolName: string, toolInput: Record<string, unknown>): void {
    this.stopThinking()
    const icon = TOOL_ICONS[toolName] || '🔧'
    const detail = this.formatDetail(toolName, toolInput)

    console.log(chalk.dim('  ⏺ Calling tools...'))
    console.log(chalk.gray('┌─ ') + chalk.bold.cyan(`${icon} ${toolName}`) + chalk.gray(' ─'))
    console.log(chalk.gray('│ ') + detail)
    console.log(chalk.gray('└─' + '─'.repeat(40)))
  }

  /** 显示工具调用结果（单行摘要） */
  showToolResult(
    toolName: string,
    result: string,
    elapsedMs: number,
    isError: boolean = false,
  ): void {
    result = result ?? ''
    this.stopThinking()
    const icon = TOOL_ICONS[toolName] || '🔧'
    const statusIcon = isError ? chalk.red('✗') : chalk.green('✓')
    const elapsed = chalk.dim.cyan(`${elapsedMs.toFixed(0)}ms`)

    const preview = result.slice(0, 200).replace(/\n/g, ' ')
    const truncated = result.length > 200 ? '...' : ''

    let line = `  ⎯ ${chalk.dim(icon)} ${chalk.bold.cyan(toolName)} ${statusIcon} ${elapsed}`
    if (isError) {
      line += `  ${chalk.red(preview + truncated)}`
    }
    console.log(line)
    this.startThinking()
  }

  /** 启动 thinking spinner（幂等） */
  startThinking(): void {
    if (!this.activeSpinner) {
      this.activeSpinner = new Spinner('Thinking...')
      this.activeSpinner.start()
    }
  }

  /** 停止 thinking spinner（幂等） */
  stopThinking(): void {
    if (this.activeSpinner) {
      this.activeSpinner.stop()
      this.activeSpinner = null
    }
  }

  /** 清理资源 */
  cleanup(): void {
    this.stopThinking()
  }

  /** 格式化工具详情行 */
  private formatDetail(toolName: string, toolInput: Record<string, unknown>): string {
    if (!toolInput || typeof toolInput !== 'object') {
      return chalk.dim(JSON.stringify(toolInput).slice(0, 150))
    }

    switch (toolName) {
      case 'bash':
        return chalk.white.bgGray(` $ ${toolInput.command || ''} `)
      case 'file_read':
        return `📄 Reading ${toolInput.path || ''}`
      case 'file_write':
        return `✏️ Writing ${toolInput.path || ''} (${String(toolInput.content ?? '').split('\n').length} lines)`
      case 'file_edit':
        return `📝 Editing ${toolInput.path || ''}`
      case 'web_search':
        return `🔍 Searching: ${toolInput.query || ''}`
      default:
        return chalk.dim(JSON.stringify(toolInput).slice(0, 150))
    }
  }
}
